subscription: Initial subscription GUI bits

M4rtinK opened the pull request:
The first commit adds the full Glade file "XML blob" providing all the spoke widget layout similar to what is currently in RHEL 8.2.

The second commit adds a couple helper functions in a separate file for easy unit testing. It is expected more will be added there later.

The third commit adds the main Subscription spoke structure. This is mostly boilerplate for interacting with the many controls on the Subscription spoke, with many of the logic methods intentionally left out for now. It is expected followup PRs of much more manageable size will add these parts. See the commit message for list of what has been left out in detail.

pep8speaks commented:
Hello @M4rtinK! Thanks for opening this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:



* In the file [`pyanaconda/ui/gui/spokes/subscription.py`](https://github.com/rhinstaller/anaconda/blob/c4ef8e40d905a97a3cb928f8e309adf81310882f/pyanaconda/ui/gui/spokes/subscription.py):

> [Line 148:5](https://github.com/rhinstaller/anaconda/blob/c4ef8e40d905a97a3cb928f8e309adf81310882f/pyanaconda/ui/gui/spokes/subscription.py#L148): [E303](https://duckduckgo.com/?q=pep8%20E303) too many blank lines (2)
> [Line 479:100](https://github.com/rhinstaller/anaconda/blob/c4ef8e40d905a97a3cb928f8e309adf81310882f/pyanaconda/ui/gui/spokes/subscription.py#L479): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (101 > 99 characters)
> [Line 498:9](https://github.com/rhinstaller/anaconda/blob/c4ef8e40d905a97a3cb928f8e309adf81310882f/pyanaconda/ui/gui/spokes/subscription.py#L498): [E266](https://duckduckgo.com/?q=pep8%20E266) too many leading '#' for block comment
> [Line 520:100](https://github.com/rhinstaller/anaconda/blob/c4ef8e40d905a97a3cb928f8e309adf81310882f/pyanaconda/ui/gui/spokes/subscription.py#L520): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (100 > 99 characters)
> [Line 522:100](https://github.com/rhinstaller/anaconda/blob/c4ef8e40d905a97a3cb928f8e309adf81310882f/pyanaconda/ui/gui/spokes/subscription.py#L522): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (102 > 99 characters)
> [Line 538:100](https://github.com/rhinstaller/anaconda/blob/c4ef8e40d905a97a3cb928f8e309adf81310882f/pyanaconda/ui/gui/spokes/subscription.py#L538): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (100 > 99 characters)
> [Line 539:100](https://github.com/rhinstaller/anaconda/blob/c4ef8e40d905a97a3cb928f8e309adf81310882f/pyanaconda/ui/gui/spokes/subscription.py#L539): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (100 > 99 characters)
> [Line 543:100](https://github.com/rhinstaller/anaconda/blob/c4ef8e40d905a97a3cb928f8e309adf81310882f/pyanaconda/ui/gui/spokes/subscription.py#L543): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (106 > 99 characters)
> [Line 544:100](https://github.com/rhinstaller/anaconda/blob/c4ef8e40d905a97a3cb928f8e309adf81310882f/pyanaconda/ui/gui/spokes/subscription.py#L544): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (106 > 99 characters)
> [Line 545:100](https://github.com/rhinstaller/anaconda/blob/c4ef8e40d905a97a3cb928f8e309adf81310882f/pyanaconda/ui/gui/spokes/subscription.py#L545): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (100 > 99 characters)
> [Line 553:9](https://github.com/rhinstaller/anaconda/blob/c4ef8e40d905a97a3cb928f8e309adf81310882f/pyanaconda/ui/gui/spokes/subscription.py#L553): [E266](https://duckduckgo.com/?q=pep8%20E266) too many leading '#' for block comment
> [Line 563:100](https://github.com/rhinstaller/anaconda/blob/c4ef8e40d905a97a3cb928f8e309adf81310882f/pyanaconda/ui/gui/spokes/subscription.py#L563): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (100 > 99 characters)
> [Line 571:9](https://github.com/rhinstaller/anaconda/blob/c4ef8e40d905a97a3cb928f8e309adf81310882f/pyanaconda/ui/gui/spokes/subscription.py#L571): [E303](https://duckduckgo.com/?q=pep8%20E303) too many blank lines (2)
> [Line 642:9](https://github.com/rhinstaller/anaconda/blob/c4ef8e40d905a97a3cb928f8e309adf81310882f/pyanaconda/ui/gui/spokes/subscription.py#L642): [E303](https://duckduckgo.com/?q=pep8%20E303) too many blank lines (2)
> [Line 653:9](https://github.com/rhinstaller/anaconda/blob/c4ef8e40d905a97a3cb928f8e309adf81310882f/pyanaconda/ui/gui/spokes/subscription.py#L653): [E265](https://duckduckgo.com/?q=pep8%20E265) block comment should start with '# '

* In the file [`tests/nosetests/pyanaconda_tests/subscription_spoke_tests.py`](https://github.com/rhinstaller/anaconda/blob/c4ef8e40d905a97a3cb928f8e309adf81310882f/tests/nosetests/pyanaconda_tests/subscription_spoke_tests.py):

> [Line 25:1](https://github.com/rhinstaller/anaconda/blob/c4ef8e40d905a97a3cb928f8e309adf81310882f/tests/nosetests/pyanaconda_tests/subscription_spoke_tests.py#L25): [E302](https://duckduckgo.com/?q=pep8%20E302) expected 2 blank lines, found 1

M4rtinK commented:
For some context, this is how the spoke currently looks like on RHEL 8.2:

![rhel82_rhsm_hub](https://user-images.githubusercontent.com/829558/81294237-e326e200-906e-11ea-8923-d0f290464997.png)
spoke status on the Hub

![rhel82_rhsm_not_registered](https://user-images.githubusercontent.com/829558/81294270-ede17700-906e-11ea-8bea-b7e633781380.png)
not yet registered with some fields filled in

![rhel82_rhsm_registered](https://user-images.githubusercontent.com/829558/81294331-0356a100-906f-11ea-9883-a76430ff488b.png)
registered & subscription attached (there could be multiple subscriptions attached, so the subscription delegates are inside a scrollable list view)

M4rtinK commented:
jenkins, test this please

poncovka commented on pyanaconda/ui/gui/spokes/subscription.py:
```diff
@@ -0,0 +1,750 @@
+
+    @property
+    def http_proxy_username(self):
+        return self._http_proxy_username_entry.get_text()
```
I am not a big fan of this approach. Instead of keeping the data in UI elements, I would prefer to keep the data in the subscription request. These properties could be replaced with signal handlers that would update the request with a new value of the changed UI element. Then you can always use the request to check the current values.

This works pretty well in the custom partitioning spoke (see https://github.com/rhinstaller/anaconda/blob/master/pyanaconda/ui/gui/spokes/custom_storage.py#L1563). I use the request to log the current state or to check if something has changed.

However, it might be a too big change that could complicate next backports, so it is up to you.

M4rtinK replied:
I think it is a good idea - this always seemed cumbersome, but with the DBus module infrastructure on RHEL 8.2 & the available timeframe I was not able to come up with something better.

I like the idea & if its known to work elsewhere I think I can make it work here as well.

poncovka commented on pyanaconda/ui/gui/spokes/subscription.py:
```diff
@@ -0,0 +1,750 @@
+    def refresh(self):
+        log.debug("Subscription GUI: refresh() running")
+        # update subscription request from the Subscription module
+        self._subscription_request = self._get_subscription_request()
```
This method should set up the UI elements from the current request.

M4rtinK replied:
Added a content refresh method called from here.

poncovka commented on pyanaconda/ui/gui/spokes/subscription.py:
```diff
@@ -0,0 +1,750 @@
+        self._unregister_revealer = self.builder.get_object("unregister_revealer")
+        self._unregister_button = self.builder.get_object("unregister_button")
+
+        # setup spoke state based on Subscription DBus module state
```
This part should be in the refresh method.

M4rtinK replied:
Split into a content refresh method.

M4rtinK commented on pyanaconda/ui/gui/spokes/subscription.py:
```diff
@@ -0,0 +1,750 @@
+        return is_module_available(SUBSCRIPTION)
+
+    def __init__(self, *args):
+        NormalSpoke.__init__(self, *args)
```
Fixed.

M4rtinK commented:
**PR updated:**
- content of the spoke is now always set based on DBus module state on spoke entry
- subscription request and system purpose data DBus structure mirrors are now updated by signal handlers when content of a field changes due to user interaction
- added a couple changed signals to the glade file, so it is no longer 100% identical to the RHEL 8.2
- DBus structure mirrors are set to the module on `apply()` or (in the future) before a subscription attempt
- support for password placeholders added (would have to be set by another client via DBus as kickstart only supports org & key authentication

**Known issues:**
- the visibility checkboxes actually have no effect on the values being set or not to the DBus module, a reasonable implementation if this needs to be added that keeps expected user behavior from 8.2 (fields are not cleared when you just toggle the checkbox) and does not complicate the code too much
- handling of the Insights default (off by default for kickstart installs, on by default for interactive GUI installs) is ugly & cold be done better if Subscription module could check if installation is a kickstart driven installation

poncovka commented on pyanaconda/ui/gui/spokes/subscription.py:
```diff
@@ -0,0 +1,769 @@
+        # If we have a kickstart, use the value from the module, otherwise enable Insights
+        # as that's the default for interactive installations.
+        if flags.automatedInstall:
+            self._insights_checkbox.set_active(self._subscription_module.InsightsEnabled)
```
This is already set up in `_update_spoke_state`. I think you should set the DBus property `InsightsEnabled` to `True` for interactive installations before you call `_update_spoke_state` (ideally at the beginning of the `initialize` or `__init__` method).

M4rtinK replied:
Nice solution, let's do that! :)

jkonecny12 commented on pyanaconda/ui/gui/spokes/lib/subscription.py:
```diff
@@ -0,0 +1,92 @@
+    :param str user_provided_value: a value provided by user
+    :param list valid_values: a list of valid values
+    :returns: list of values with one value preselected
+    :rtype: list of (str, str, bool) tuples in (<value>, <display string>, <is_preselected>) format
```
What about using named tuple. It's really inconvenient to read code with three tuple because you don't know what is what and [2] doesn't really help here.

Or mark this as private method and leave it as it is. I don't mind if it is not an API.

M4rtinK replied:
Yep, named tuple looks better, let's use it. :)

jkonecny12 commented on pyanaconda/ui/gui/spokes/subscription.py:
```diff
@@ -0,0 +1,769 @@
+        if entered_text:
+            proxy_obj = ProxyString(url=entered_text)
+            hostname = proxy_obj.host
+            port = proxy_obj.port
```
I think you want to remove this line. It's `-1` in case of not set and if set then the value is changed in the if below.

M4rtinK replied:
Yep, that's redundant, lets remove it.

jkonecny12 commented on pyanaconda/ui/gui/spokes/subscription.py:
```diff
@@ -0,0 +1,769 @@
+        self._subscription_request = self._get_subscription_request()
+        self._system_purpose_data = self._get_system_purpose_data()
+
+        # authentication method
+        # - SubscriptionRequest always has type set
+        # - username + password is the default
+        # For the related password and activation keys entry holding sensitive data
+        # we need to reconcile the data held in the spoke from previous entry with
+        # data set in the DBus module previously:
+        # - data in module and entry empty -> set placeholder
+        # - data in module and entry populated -> keep text in entry,
+        #   we assume it is the same as what is in module
+        # - no data in module and entry populated -> clear entry & any placeholders
+        #   (data cleared over DBus API)
+        # - no data in module and entry empty -> do nothing
+        if self.subscription_request.type == SUBSCRIPTION_REQUEST_TYPE_USERNAME_PASSWORD:
+            self.authentication_method = AuthenticationMethod.USERNAME_PASSWORD
+            self.username_entry.set_text(self.subscription_request.account_username)
+            set_in_entry = bool(self._password_entry.get_text())
+            set_in_module = self.subscription_request.account_password.type == SECRET_TYPE_HIDDEN
+            if set_in_module:
+                if not set_in_entry:
+                    self.enable_password_placeholder(True)
+            else:
+                self._password_entry.set_text("")
+                self.enable_password_placeholder(False)
+        elif self.subscription_request.type == SUBSCRIPTION_REQUEST_TYPE_ORG_KEY:
+            self.authentication_method = AuthenticationMethod.ORG_KEY
+            self._organization_entry.set_text(self.subscription_request.organization)
+            set_in_entry = bool(self._activation_key_entry.get_text())
+            set_in_module = self.subscription_request.activation_keys.type == SECRET_TYPE_HIDDEN
+            if set_in_module:
+                if not set_in_entry:
+                    self.enable_activation_key_placeholder(True)
+            else:
+                self._activation_key_entry.set_text("")
+                self.enable_activation_key_placeholder(False)
```
This could be separated to a private method with the comment having as docstring.

M4rtinK replied:
Yeah, that looks a bit better.

jkonecny12 commented on pyanaconda/ui/gui/spokes/subscription.py:
```diff
@@ -0,0 +1,769 @@
+        # Insights
+        self._insights_checkbox.set_active(self._subscription_module.InsightsEnabled)
+
+        # set HTTP proxy data
+        proxy_hostname = self.subscription_request.server_proxy_hostname
+        proxy_port = self.subscription_request.server_proxy_port
+        proxy_port_set = proxy_port >= 0
+        proxy_username = self.subscription_request.server_proxy_user
+        proxy_password_secret = self.subscription_request.server_proxy_password
+        proxy_password_set = proxy_password_secret.type == SECRET_TYPE_HIDDEN
+        self.http_proxy_visible = proxy_hostname or proxy_username or proxy_password_set
+        if proxy_hostname:
+            proxy_url = proxy_hostname
+            if proxy_port_set:
+                proxy_url = "{}:{}".format(proxy_url, proxy_port)
+            self._http_proxy_location_entry.set_text(proxy_url)
+        # HTTP proxy username
+        self._http_proxy_username_entry.set_text(proxy_username)
+        # HTTP proxy password
+        set_in_entry = bool(self._http_proxy_password_entry.get_text())
+        secret_type = self.subscription_request.server_proxy_password.type
+        set_in_module = secret_type == SECRET_TYPE_HIDDEN
+        if set_in_module:
+            if not set_in_entry:
+                self.enable_http_proxy_password_placeholder(True)
+        else:
+            self._http_proxy_password_entry.set_text("")
+            self.enable_http_proxy_password_placeholder(False)
```
Second part which can be moved to a private method.

M4rtinK replied:
Moved.

poncovka commented on pyanaconda/ui/gui/spokes/subscription.py:
```diff
@@ -0,0 +1,830 @@
+        # For kickstart installations we will use the value from the module, which
+        # False by default & can be set to True via the rhsm kickstart command.
+        if not flags.automatedInstall:
+            self._subscription_module.InsightsEnabled = True
```
I think that you need to call `SetInsightsEnabled`.

M4rtinK replied:
Indeed! Fixed. :)
